internal/controller/csiaddons/persistentvolumeclaim_controller.go

```diff
@@ -210,97 +215,47 @@ func (r *PersistentVolumeClaimReconciler) checkDriverSupportCapability(
-		// pool. This can put the controller in a requeue loop. Hence we are
-		// reading the driver name from the namespace annotation and checking if
-		// the driver is registered in the connection pool and if not we are not
-		// requeuing the request.
-		// Depending on requeue value, it will return ErrorConnNotFoundRequeueNeeded.
-		switch annotationKey {
-		case krcJobScheduleTimeAnnotation:
-			requeue, keyRotationSupported := r.checkDriverSupportCapability(logger, ns.Annotations, driverName, keyRotationOp)
-			if keyRotationSupported {
-				return schedule, nil
-			}
-			if requeue {
-				return "", ErrConnNotFoundRequeueNeeded
-			}
-		case rsCronJobScheduleTimeAnnotation:
-			requeue, supportReclaimspace := r.checkDriverSupportCapability(logger, ns.Annotations, driverName, relciamSpaceOp)
-			if supportReclaimspace {
-				// if driver supports space reclamation,
-				// return schedule from ns annotation.
-				return schedule, nil
-			}
-			if requeue {
-				// The request needs to be requeued for checking
-				// driver support again.
-				return "", ErrConnNotFoundRequeueNeeded
-			}
-		default:
-			logger.Info("Unknown annotation key", "AnnotationKey", annotationKey)
-			return "", fmt.Errorf("unknown annotation key: %s", annotationKey)
+	switch r.SchedulePrecedence {
+	case util.SchedulePVCFirst:
+		// Check on PVC
+		if schedule = r.getScheduleFromPVC(pvc, logger, annotationKey); schedule != "" {
+			return schedule, nil
 		}
-	}
 
-	// For static provisioned PVs, StorageClassName is nil or empty.
-	if pvc.Spec.StorageClassName == nil || len(*pvc.Spec.StorageClassName) == 0 {
-		logger.Info("StorageClassName is empty")
-		return "", ErrScheduleNotFound
-	}
-	storageClassName := *pvc.Spec.StorageClassName
-
-	// check for storageclass schedule annotation.
-	sc := &storagev1.StorageClass{}
-	err = r.Client.Get(ctx, types.NamespacedName{Name: storageClassName}, sc)
-	if err != nil {
-		if apierrors.IsNotFound(err) {
-			logger.Error(err, "StorageClass not found", "StorageClass", storageClassName)
-			return "", ErrScheduleNotFound
+		// Check on NS, might get ErrConnNotFoundRequeueNeeded
+		// If so, return the error
+		if schedule, err = r.getScheduleFromNS(ctx, pvc, logger, driverName, annotationKey); schedule != "" {
+			return schedule, nil
+		}
+		if errors.Is(err, ErrConnNotFoundRequeueNeeded) {
+			return "", err
 		}
 
-		logger.Error(err, "Failed to get StorageClass", "StorageClass", storageClassName)
-		return "", err
-	}
-	schedule, scheduleFound = getScheduleFromAnnotation(annotationKey, logger, sc.Annotations)
-	if scheduleFound {
-		return schedule, nil
+		// Check SC
+		if schedule = r.getScheduleFromSC(ctx, pvc, logger, annotationKey); schedule != "" {
+			return schedule, nil
+		}
+	case util.ScheduleSCFirst:
+		// DS flag, read only from the SC
+		if schedule = r.getScheduleFromSC(ctx, pvc, logger, annotationKey); schedule != "" {
+			return schedule, nil
+		}
```

Comment:
The option should be SC-only instead pvc-first/sc-first? 

This would make is a lot simpler.

@black-dragon74 @Madhu-1  ?

Reply:
I agree. Having one mode of operation is always better... But it was the general consensus to preserve the existing (current) functionality.

Reply:
> I agree. Having one mode of operation is always better... But it was the general consensus to preserve the existing (current) functionality.

I meant SC-only option which toggles between pvc>ns>sc and only sc annotation.

Reply:
Ah, right. Makes sense :)

Reply:
>`// DS flag, read only from the SC`

This comment should be present. The option can be used by all users.

Reply:
IG you meant shouldn't be present? Removed it.